stl/inc/algorithm

```diff
@@ -665,13 +673,27 @@ _NODISCARD _FwdIt partition_point(_FwdIt _First, _FwdIt _Last, _Pr _Pred) {
 // FUNCTION TEMPLATE _Equal_rev_pred_unchecked
 #if _HAS_IF_CONSTEXPR
 template <class _InIt1, class _InIt2, class _Pr>
-bool _Equal_rev_pred_unchecked(_InIt1 _First1, _InIt2 _First2, const _InIt2 _Last2, _Pr _Pred) {
+_NODISCARD _CONSTEXPR20_ICE bool _Equal_rev_pred_unchecked(
+    _InIt1 _First1, _InIt2 _First2, const _InIt2 _Last2, _Pr _Pred) {
     // compare [_First1, ...) to [_First2, _Last2) using _Pred
     if constexpr (decltype(_Equal_memcmp_is_safe(_First1, _First2, _Pred))::value) {
-        const auto _First1_ch = reinterpret_cast<const char*>(_First1);
-        const auto _First2_ch = reinterpret_cast<const char*>(_First2);
-        const auto _Count     = static_cast<size_t>(reinterpret_cast<const char*>(_Last2) - _First2_ch);
-        return _CSTD memcmp(_First1_ch, _First2_ch, _Count) == 0;
+#ifdef __cpp_lib_is_constant_evaluated
+        if (_STD is_constant_evaluated()) {
+            for (; _First2 != _Last2; ++_First1, (void) ++_First2) {
+                if (!_Pred(*_First1, *_First2)) {
+                    return false;
+                }
+            }
+
+            return true;
+        } else
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            const auto _First1_ch = reinterpret_cast<const char*>(_First1);
+            const auto _First2_ch = reinterpret_cast<const char*>(_First2);
+            const auto _Count     = static_cast<size_t>(reinterpret_cast<const char*>(_Last2) - _First2_ch);
+            return _CSTD memcmp(_First1_ch, _First2_ch, _Count) == 0;
+        }

```

Comment:
Can't we do better with `__builtin_memcmp`? At least for non-CUDA?

Reply:
There's no point to do that -- `__builtin_memcmp` is the same as `memcmp` at runtime, and the reinterpret_casts before already exclude this block from constexpr processing.

Reply:
The point is throughput: we don't need to `if (is_constant_evaluated())` with `__builtin_memcmp` since it's valid at both compile time and at runtime. The body becomes:
```c++
if constexpr (decltype(_Equal_memcmp_is_safe(_First1, _First2, _Pred))::value) {
    return 0 == __builtin_memcmp(_First1, _First2, (_Last2 - _First2) * sizeof(*_First2));
} else {
    // as before
}
```

Reply:
(1) I believe that violates the contract for `__builtin_memcmp` at constexpr time for us; I believe it accepts chars only. At least, we are guarding it that way in the only place we currently use it.

(2) That gives up on the optimization we're getting from the reinterpret_cast -- it'll introduce the divide + multiply from the pointer math.

(3) The throughput difference of parsing this loop is so So SO far in the noise I don't know why we're even worrying about it.

Reply:
(1) Apparently so: VSO-714690.

(2) "Divide + multiply" what? [Oh dear](https://godbolt.org/z/8poAr3). Nevermind.

(3) Size of source code is also a metric. It would be nice if we could make this function body a dozen lines shorter.

Reply:
RE: (2) @manbearian said they were looking at such problems once upon a time